### ISSUE TITLE
[Question] How to calculate rewards only at the end of an episode?

### ISSUE THREAD

**Issue by H-Hisamichi** [isaac-sim/IsaacLab#4067]
### Question

Hello everyone,

I am working on a DRL controller for a jumping robot. I have a specific reward that I want to calculate only at the end of an episode to evaluate if the robot reached the target height.

Does IsaacLab have an option to change the timing of the reward calculation?

Thanks.

### Build Info

- Isaac Lab Version: [2.2.0]
- Isaac Sim Version: [5.0.0]

**Comment by RandomOakForest** [comment]
Thank you for posting this. Yes, you can implement a reward that is only given at the end of an episode, but Isaac Lab does not have a global “compute rewards only at terminal step” switch. Instead, you implement this as a normal reward term and gate its value so it is non‑zero only for environments that are terminating at the current step. Here are a few notes to consider.

## How rewards are computed

Isaac Lab’s reward manager calls all configured reward terms every simulation step and sums their weighted outputs into a per‑step reward tensor.  Each term is just a function mapping the current environment state (and possibly commands) to a tensor of shape $`(\text{num\_envs},)`$.  The manager also keeps episodic sums of each term internally, but the policy update still uses the per‑step rewards it computes.[^1][^2][^3]

## Accessing “done”/termination info

In a manager‑based RL environment, term functions receive the environment instance and can read termination buffers like:

- `env.termination_manager.terminated` (per‑env boolean for true terminal).
- `env.termination_manager.time_outs` (per‑env boolean for time‑limit terminations).

These are the same signals that `_get_dones()` uses to mark which envs reset.  You can use these to detect “this step is the last step of the episode for env i”.[^4][^5]

## Implementing an end-of-episode reward term

You implement your “jump height success” as a reward term that returns zero at all steps, except on the step where an environment terminates and meets your height condition. Conceptually:

- Compute current COM height or foot height from the robot asset.
- Check if that env is terminating now (e.g., `terminated | time_out`).
- If terminating and height ≥ target, output a positive reward; otherwise output zero.

In a custom reward function file (similar to other terms in `isaaclab.envs.mdp.rewards` ):[^6][^3]

```python
import torch
from isaaclab.envs import ManagerBasedRLEnv
from isaaclab.scene import SceneEntityCfg

def jump_success_terminal(
    env: ManagerBasedRLEnv,
    target_height: float,
    asset_cfg: SceneEntityCfg = SceneEntityCfg("robot"),
) -> torch.Tensor:
    robot = env.scene[asset_cfg.name]
    # Example: use base COM z position
    base_height = robot.data.root_com_pos_w[:, 2]

    # Detect terminal step (include or exclude timeouts as desired)
    terminated = env.termination_manager.terminated
    time_outs = env.termination_manager.time_outs
    is_terminal_step = terminated | time_outs  # or just `terminated` if you prefer

    # End-of-episode sparse reward
    success = (base_height >= target_height) & is_terminal_step
    reward = success.float()  # e.g. 1.0 for success, 0 otherwise

    return reward
```

Then add it to your environment config’s `RewardTermCfg` list with some weight, like:

```python
from isaaclab.managers import RewardTermCfg
from isaaclab.envs import ManagerBasedRLEnvCfg

class MyJumpEnvCfg(ManagerBasedRLEnvCfg):
    # ...
    class rewards(ManagerBasedRLEnvCfg.rewards):
        jump_success_terminal = RewardTermCfg(
            func="my_rewards.jump_success_terminal",
            weight=10.0,
            params={"target_height": 0.8},
        )
```

This way:

- The term is evaluated every step, but it returns non‑zero only on terminal steps.
- The resulting signal is a sparse, end‑of‑episode bonus that integrates cleanly into Isaac Lab’s normal reward pipeline.[^3][^6]


[^1]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.managers.html

[^2]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.managers.html#reward-manager

[^3]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.envs.mdp.html

[^4]: https://isaac-sim.github.io/IsaacLab/main/source/setup/walkthrough/technical_env_design.html

[^5]: https://isaac-sim.github.io/IsaacLab/main/source/tutorials/03_envs/create_manager_rl_env.html

[^6]: https://isaac-sim.github.io/IsaacLab/main/_modules/isaaclab/envs/mdp/rewards.html

[^7]: https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/training_guide.html

[^8]: https://isaac-sim.github.io/IsaacLab/main/source/setup/walkthrough/training_jetbot_reward_exploration.html

[^9]: https://isaac-sim.github.io/IsaacLab/main/source/how-to/curriculums.html

[^10]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab_rl/isaaclab_rl.html

[^11]: https://github.com/isaac-sim/IsaacLab/issues/4067

[^12]: https://arxiv.org/html/2511.19355v1

[^13]: https://docs.nvidia.com/learning/physical-ai/getting-started-with-isaac-lab/latest/train-your-second-robot-with-isaac-lab/06-custom-reward-functions-and-hyperparameters.html

[^14]: https://docs.nvidia.com/learning/physical-ai/getting-started-with-isaac-lab/latest/train-your-first-robot-with-isaac-lab/01-what-is-reinforcement-learning.html

[^15]: https://github.com/isaac-sim/IsaacLab/issues/2079

[^16]: https://www.youtube.com/watch?v=o_KokSzb8NI

[^17]: https://www.reddit.com/r/reinforcementlearning/comments/1jmzyfp/master_thesis_reinforcement_learning_of_humanoid/

[^18]: https://github.com/isaac-sim/IsaacLab/issues/825

[^19]: https://forums.developer.nvidia.com/t/isaac-lab-rl/307090

[^20]: https://www.reddit.com/r/reinforcementlearning/comments/18eeutp/how_to_upgrade_ai_with_rewards_only_given_at_the/

**Comment by H-Hisamichi** [comment]
Hi @RandomOakForest,

Thank you for the advice! I was able to implement the sparse reward based on your suggestion.

However, since I want to train the full sequence from takeoff to landing using DRL, I created a modified reward called "jump_success_terminal_latched". In this function, I create a custom buffer directly within "env" to latch the "target height reached" flag during the episode.

Is this method of creating a custom buffer appropriate? It seems to be working correctly for now, but I would appreciate your advice if there is a more standard or better way to handle such state buffers in Isaac Lab.

Thanks!

```python
def jump_success_terminal_latched(
    env: ManagerBasedRLEnv, command_name: str, asset_cfg: SceneEntityCfg = SceneEntityCfg("robot"),
) -> torch.Tensor:
    """
    A reward that returns 1.0 at the end of an episode if the target height is reached at least once during the episode.
    """
    robot = env.scene[asset_cfg.name]
    current_height = robot.data.root_com_pos_w[:, 2]
    target_height = env.command_manager.get_command(command_name)[:, 2]
    # initialize the buffer for maintaining the state
    if not hasattr(env, "_jump_success_latch"):
        env._jump_success_latch = torch.zeros(env.num_envs, dtype=torch.bool, device=env.device)
    
    # reset flag to False at episode start
    reset_env_ids = (env.episode_length_buf == 0).nonzero(as_tuple=False).flatten()
    if len(reset_env_ids) > 0:
        env._jump_success_latch[reset_env_ids] = False
    
    # update flags
    is_above_target = current_height >= target_height
    env._jump_success_latch = env._jump_success_latch | is_above_target

    # detect terminal step (include or exclude timeouts as desired)
    terminated = env.termination_manager.terminated
    time_outs = env.termination_manager.time_outs
    is_terminal_step = terminated | time_outs  # or just `terminated` if you prefer
    return (env._jump_success_latch & is_terminal_step).float()
```

**Comment by RandomOakForest** [comment]
Yes, latching a “target height reached” flag in a custom buffer on the env is an appropriate and common pattern in Isaac Lab, as long as you respect the reset semantics and device placement.
